cpp/src/filling/fill.cu

```diff
@@ -0,0 +1,147 @@
+  operator()(cudf::size_type source_begin, cudf::size_type source_end,
+             cudf::size_type target_begin, cudaStream_t stream = 0) {
+    using ScalarType = cudf::experimental::scalar_type_t<T>;
+#if 1
+    // TODO: temporary till the const issue in cudf::scalar's value() is fixed.
+    auto p_scalar =
+      const_cast<ScalarType*>(static_cast<ScalarType const*>(this->p_value));
+#else
+    auto p_scalar = static_cast<ScalarType const*>(this->p_value);
+#endif
+    T value = p_scalar->value(stream);
+    bool is_valid = p_scalar->is_valid();
+    cudf::experimental::detail::copy_range(
+      thrust::make_constant_iterator(value),
+      thrust::make_constant_iterator(is_valid),
+      target, target_begin, target_begin + (source_end - source_begin),
+      stream);
+  }
+
+  template <typename T>
+  std::enable_if_t<not cudf::is_fixed_width<T>(), void>
+  operator()(cudf::size_type source_begin, cudf::size_type source_end,
+             cudf::size_type target_begin, cudaStream_t stream = 0) {
+    CUDF_FAIL("in-place fill does not work for variable width types.");
+  }
+};
+
+}  // namespace
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+void fill(mutable_column_view& destination,
+          size_type begin,
+          size_type end,
+          scalar const& value,
+          cudaStream_t stream) {
+  CUDF_EXPECTS(cudf::is_fixed_width(destination.type()) == true,
+               "In-place fill does not support variable-sized types.");
+  CUDF_EXPECTS((begin >= 0) &&
+               (begin <= end) &&
+               (begin < destination.size()) &&
+               (end <= destination.size()),
+               "Range is out of bounds.");
+  CUDF_EXPECTS((destination.nullable() == true) || (value.is_valid() == true),
+               "destination should be nullable or value should be non-null.");
+  CUDF_EXPECTS(destination.type() == value.type(), "Data type mismatch.");
+
+  if (end != begin) {  // otherwise no-op
+    cudf::experimental::type_dispatcher(
+      destination.type(),
+      inplace_fill_range_dispatch{&value, destination},
+      0, end - begin, begin, stream);
+  }
+
+  return;
+}
+
+std::unique_ptr<column> fill(column_view const& input,
+                             size_type begin,
+                             size_type end,
+                             scalar const& value,
+                             cudaStream_t stream,
+                             rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS(cudf::is_fixed_width(input.type()) == true,
+               "Variable-sized types are not supported yet.");
+  CUDF_EXPECTS((begin >= 0) &&
+               (begin <= end) &&
+               (begin < input.size()) &&
+               (end <= input.size()),
+               "Range is out of bounds.");
+  CUDF_EXPECTS(input.type() == value.type(), "Data type mismatch.");
+
+  auto p_ret = std::make_unique<column>(input, stream, mr);
+  if (!p_ret->nullable() && !value.is_valid()) {
```

Comment:
I haven't thought very much about such a case, in that case, I think both are legitimate (allocating or not allocating null_mask), but if cuDF code maintainers (@jrhemstad @harrism) have a preference, then, I will implement that.

Reply:
I don't follow. Why is `end != begin` a noop?

Reply:
@harrism If the copy_range is 0 (if end == begin), copy_range becomes a no-op.

Reply:
I guess (

> because end != begin

) in @cwharris's comment is a typo, it guess actual intention is (end == begin).

Reply:
So, the question is

in the out-of-place version of the fill funciton,
```
std::unique_ptr<column> fill(column_view const& input,
                             size_type begin,
                             size_type end,
                             scalar const& value,
                             cudaStream_t stream,
                             rmm::mr::device_memory_resource* mr) {
```

If a user

1) asks to fill with a null value (_value_.is_valid() is false)
2) but the copy range is empty (begin == end),
3) and _target_ is not nullable

then, should the returned column be nullable or not? I guess if we weigh on user's intention to fill with a null value, then, the returned column should be nullable(). If we focus on the empty copy range (so this might be considered as a no-op), then, we should not make the returned column nullable I guess.

Reply:
I see at least two good options, because IMHO consistency is more important than efficiency:
- 1. Always return a null mask in the column if `fill_value` is null. The user can easily choose to not invoke the function based on `begin == end` arguments.
- 2. Allow the user to specify behavior via optional argument such as `enforce_fill_value_nullability = true`.

Reply:
I personally prefer the option 1 (and this is the current implementation if I am not mistaken). Adding _enforce_fill_value_nullability_ makes the function signature a bit too complicated with only very little benefit.

Reply:
I agree with option 1.